helper/scaleutils/nomad.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// If we have not been able to identify any nodes and get their remote
+	// provider ID we cannot continue.
+	if len(nodeIDMap) == 0 {
+		return nil, errors.New("failed to identify nodes for removal")
+	}
+
+	if err := si.drainNodes(ctx, req.DrainDeadline, nodeIDMap); err != nil {
+		return nil, err
+	}
+
+	return nodeIDMap, nil
+}
+
+// identifyTargets filters the current Nomad cluster node list and then sorts
+// and selects nodes for removal based on the specified strategy. It is
+// possible the list does not contain any many nodes as requested do to the
+// limited number available after filtering.
+func (si *ScaleIn) identifyTargets(num int, ident *PoolIdentifier, strategy NodeIDStrategy) ([]*api.NodeListStub, error) {
+
+	// Pull a current list of Nomad nodes from the API.
+	nodes, _, err := si.nomad.Nodes().List(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list Nomad nodes from API: %v", err)
+	}
+
+	// Filter the node list depending on what pool identifier we are using.
+	si.log.Debug("filtering node list", "filter", ident.IdentifierKey, "value", ident.Value)
+
+	switch ident.IdentifierKey {
+	case IdentifierKeyClass:
+		si.log.Debug("filtering node list by class", "class", ident.Value)
+		nodes = filterByClass(nodes, ident.Value)
+	default:
+		return nil, fmt.Errorf("unsupported node pool identifier: %q", ident.IdentifierKey)
+	}
+
+	// Ensure we have not filtered out all the available nodes.
+	if len(nodes) == 0 {
+		return nil, fmt.Errorf("no nodes unfiltered for %s with value %s", ident.IdentifierKey, ident.Value)
+	}
+
+	// Identify the strategy we are using to pick nodes for scale in and
+	// perform our list sorting.
+	switch strategy {
+	case IDStrategyNewestCreateIndex:
+	default:
+		return nil, fmt.Errorf("unsupported scale in node identification strategy: %q", strategy)
+	}
+
+	// If the caller has requested more nodes than we have available once
+	// filtered, adjust the value. This shouldn't cause the whole scaling
+	// action to fail, but we should warn.
+	if num > len(nodes) {
+		si.log.Warn("can only identify portion of requested nodes for removal",
+			"requested", num, "available", len(nodes))
+		num = len(nodes)
+	}
+
+	var out []*api.NodeListStub
+
+	// Iterate through our filtered and sorted list of nodes, selecting the
+	// required number of nodes to scale in.
+	for i := 0; i <= num-1; i++ {
+		si.log.Debug("identified Nomad node for removal", "node_id", nodes[i].ID)
+		out = append(out, nodes[i])
+	}
+
+	return out, nil
+}
+
+func (si *ScaleIn) getRemoteIDMap(nodes []*api.NodeListStub, remoteProvider RemoteProvider) ([]NodeIDMap, error) {
+
+	var idFunc nodeIDMapFunc
+
+	// Depending on the remote provider we are targeting, identify the function
+	// we need to use.
+	switch remoteProvider {
+	case RemoteProviderAWS:
+		idFunc = awsNodeIDMap
+	default:
+		return nil, fmt.Errorf("unsupported remote provider: %q", remoteProvider)
+	}
```

Comment:
Since there seems to be a 1:1 relationship between provider and ID function, we can avoid having a `switch` here by using a `map[RemoteProvider]nodeIDMapFunc`, so this block would be something like:

```go
idFunc, ok := idFuncMap[remoteProvider]
if !ok {
  return nil, ...
}
```

and we would avoid extra `case`s when adding other cloud providers.

Reply:
Yeah thats a much nicer idea!